pkg/network/prober/prober.go

```diff
@@ -30,16 +30,49 @@ import (
 	"github.com/knative/serving/pkg/network"
 )
 
+// Prober is implemented by any value that has a Do and Offer methods,
+// which respectively are synchronous and asynchronous probe methods.
+type Prober interface {
+	// TODO(bancel): rename to something more meaningful (e.g. ProbeSync, ProbeAsync)
+	Do(ctx context.Context, target, headerValue string, pos ...ProbeOption) (bool, error)
+	Offer(ctx context.Context, target, headerValue string, callback Callback, period, timeout time.Duration) bool
+}
+
 // TransportFactory is a function which returns an HTTP transport.
 type TransportFactory func() http.RoundTripper
 
 // ProbeOption is a way for caller to modify the HTTP request before it goes out.
 type ProbeOption func(r *http.Request) *http.Request
 
+// Callback is executed when the an async probe completes.
+type Callback func(success bool, err error)
+
+// Manager manages async probes and makes sure we run concurrently only a single
+// probe for the same key.
+type Manager struct {
```

Comment:
this doesn't have to be a public name anymore, since you're accessing via interface.